### ISSUE TITLE
Stormpy-pomdp: Alternative for `--belief-exploration` and `--memorybound` CLI Parameters

### ISSUE THREAD

**Issue by Zedrichu** [moves-rwth/stormpy#309]
After an extensive search through the documentation for the `pomdp` module of `stormpy`, I am unable to enable the use of memory-less controllers through binders. In the CLI, I run the commands with:
```bash
storm-pomdp --prism ... --buildfull --belief-exploration --memorybound 1 --exact
```
From my understanding, the `--buildfull` parameter can be substituted by BuilderOptions for all reward models and all labels. Similarly, I detected the `BeliefExplorationModelCheckerDouble` that seems to run instead of `--belief-exploration`. Finally, I cannot substitute the `--memorybound 1` parameter, as it is not enabled as a model checker option for belief explorations. 

Can I somehow use the `PomdpMemory` class to set these options for the model checker? Is there another alternative to this CLI option in `storm-pomdp`?

**Comment by AlexBork** [comment]
Hey Adrian,

the ``--memorybound`` parameter does not have any effect on the belief exploration. It is related to the POMDP-to-parametric-Markov-chain transformation method for synthesis of finite memory policies on POMDPs (see e.g. https://arxiv.org/abs/1710.10294).
The belief exploration inherently considers strategies with memory *on the POMDP* as the belief MDP encodes information about the history of observations in its state space. The strategies computed for the belief MDP are memoryless (i.e. have a memory of 1) *on the belief MDP*, but these then correspond to finite-memory policies *on the POMDP* (see e.g. Section 4.3 of https://arxiv.org/pdf/2305.14149).

I hope this clarifies things. Let me know if you need any elaboration, I am always happy to talk about these things :)

**Comment by Zedrichu** [comment]
Hi Alex,
Thank you for your time and explanations. I actually had one of the papers next up in my list.

According the help docs in `storm-pomdp` it seemed that this argument allowed us to control the memory size in the controller:
```
--[pomdp:]memorybound <bound> ................................. Sets the maximal number of allowed memory states (1 means memoryless schedulers).
```
If we were to use this argument alone and without a solving strategy, would it essentially achieve memory-less strategies?

As I see currently through the binder, there's no direct way of solving POMDPs (it is not as simple as parsing, building and model checking as the tutorials suggest). It would be nice to have an example about how to connect the Python binders and run a POMDP solver - and then have an explanation about what each solver does. From the library, I could also notice some BeliefTrackers and the IterativeQualitativeSearchSolver that may be relevant.

We are working with a restriction on positional strategies, hence we require memory-less policies, but finite-state may not be necessarily memory-less. Now that you mention it, it definitely makes sense that the belief MDP encodes memory-full policies. 
Since memory-less may be less performant, I can use the results from Storm only as an over-approximation for my rewards.

Is there any way to synthesise memory-less controllers for POMDPs through Storm? Would PAYNT then be of use for us (I remember there was a parameter about FSC memory size)?

**Comment by AlexBork** [comment]
Storm-POMDP does not directly support solving POMDPs under strictly memory-less strategies. Concerning the CLI, the ``--memorybound <bound>`` parameter triggers a transformation of the POMDP such that the memory of the strategy is encoded in its state space. This means that a memory-less strategy on the transformed model corresponds to a finite memory strategy with ``<bound>`` many memory nodes on the original model. If I'm not missing anything, this is currently only really used for the POMDP-to-pMC transformation.

The way Storm supports solving POMDPs under memory-less strategies is by first transfoming the POMDP to a parametric Markov chain (pMC) and then solving that using the methods for parametric models. Note that this in general considers strategies with *randomised* action-choices, so if you need action-determinism, this might not be what you are looking for.
As far as I know, the transformation is available in stormpy via ``apply_unknown_fsc`` which takes the POMDP object as input and outputs a pMC object. @sjunges might know a bit more about this.

PAYNT should indeed be useful for your purpose. They synthesise (deterministic) finite-memory controllers, and in particular, as you noted, you can restrict the memory to be 1, so you get a *deterministic* memory-less strategy.

Finally, the bindings for the belief exploration are very rudimentary and are currently very much designed with a focus on the integration with PAYNT. On the Storm side, I am currently re-working a lot of things regarding the belief exploration, and my plan is that we are then able to provide more general (and better documented :) ) bindings. Out of curiosity, would having the over-approximations you get based on belief exploration be useful for your purpose?

**Comment by TheGreatfpmK** [comment]
Hey guys,

If the use case is finding deterministic memoryless policies for POMDPs then PAYNT really is what you are looking for. Actually, if you call PAYNT with a POMDP, then it searches for memoryless policies by default. Let me know if you need any help setting that up, I can also prepare you a simple script that showcases how to call PAYNT, provided you already have a storm POMDP loaded in memory if that would be useful for you.

**Comment by Zedrichu** [comment]
We are considering both _deterministic_ and _randomised_ memory-less strategies. The context is trying to synthesise appropriate observation functions for a family of POMDPs, based on this earlier paper:
https://link.springer.com/chapter/10.1007/978-3-031-65633-0_17

Somehow for the _deterministic_ case we are already satisfied with the performance of parameter synthesis through SMT. We focus therefore on scaling the randomized strategies. I will try to implement the flow with the transformation to achieve randomised memory-less strategies. On that note, we are looking for a way to parse a Prism property into a `stormpy.logic.Formula`. The alternative is to port everything to PAYNT as we already have the PRISM models pre-built - and we will definitely try that as well. 

Regarding your final question: It is an idea that popped today after your first answer as I realised it's not in fact memory-less. We are trying to train a deep RL module for this observation function synthesis. Since positional strategies are not guaranteed to be optimal, the results based on bound FSC may be misleading as a fitness function for learning. Anyhow, they might be a better heuristic than anything else in our arsenal - hopefully the observation functions derived through memory-full strategies are correlated with the ones from memory-less. It's all hypotheses and experimental in the current stage, of course.

Thank you guys for the effort of explaining this and providing some guidance.

**Comment by Zedrichu** [comment]
@TheGreatfpmK Can PAYNT handle randomised memoryless policies on the other side?
If you have such an example for loading PAYNT, it would be very much appreciated. Currently, my code works along the following lines:
```python
program = stormpy.parse_prism_program(self.world_config.model_path, simplify=True)
property_str = f"Rmin=?[ F \"gameover\"]"
property = stormpy.parse_properties_for_prism_program(property_str, self.static_program)
program.define_constants(storm_mappings)
options = stormpy.BuilderOptions()
options.set_build_all_reward_models(True)
options.set_build_all_labels(True)
model = stormpy.build_sparse_exact_model_with_options(self.static_program,options)
checker = stormpy.pomdp.BeliefExplorationModelCheckerDouble(model, belexpl_options)
result = checker.check(property, ...)
```
If you have any tips for working more efficiently with them, let us know.

> Hey guys,
> 
> If the use case is finding deterministic memoryless policies for POMDPs then PAYNT really is what you are looking for. Actually, if you call PAYNT with a POMDP, then it searches for memoryless policies by default. Let me know if you need any help setting that up, I can also prepare you a simple script that showcases how to call PAYNT, provided you already have a storm POMDP loaded in memory if that would be useful for you.

**Comment by linusheck** [comment]
> We are considering both _deterministic_ and _randomised_ memory-less strategies. The context is trying to synthesise appropriate observation functions for a family of POMDPs, based on this earlier paper: https://link.springer.com/chapter/10.1007/978-3-031-65633-0_17

What you want to do depends a lot on your exact problem statement, but you can compute _randomized_ memoryless strategies using Storm by converting to parametric Markov chains akin to https://arxiv.org/abs/1710.10294 and then using Gradient Descent to find a feasible instantiation akin to https://arxiv.org/abs/2111.04407. PAYNT is better (or exclusively synthesizes?) deterministic strategies.

- Convert POMDP to pMC: See discussion above, I can also find more concrete instructions if you are stuck here
- https://www.stormchecker.org/documentation/usage/running-storm-on-parametric-models.html see "Feasibility with gd": Run Gradient Descent on the resulting pMC. There should also be some stormpy bindings for this

You might also be interested in this paper and its code: https://arxiv.org/pdf/2505.09518
They use both PAYNT and GD for pMCs.

**Comment by Zedrichu** [comment]
Thank you for your support. We will try to regroup and have a go at your pointers.

**Comment by TheGreatfpmK** [comment]
PAYNT synthesizes deterministic strategies. As Linus said, if you want randomized then you should run some GD approach. Anyway, here is a small script that you can use to call PAYNT synthesis:

```python
import paynt
from paynt.quotient.pomdp import PomdpQuotient
import paynt.verification.property
import paynt.synthesizer.synthesizer_ar
import stormpy

def run_paynt(model, property):
    # initialize specifcation
    properties = []
    prop = paynt.verification.property.construct_property(property[0], 0)
    properties.append(prop)
    specification = paynt.verification.property.Specification(properties)

    # create quotient
    quotient = PomdpQuotient(model, specification)

    # call synthesis
    synthesizer = paynt.synthesizer.synthesizer_ar.SynthesizerAR(quotient)
    synthesizer.synthesize(keep_optimum=True, print_stats=False)

    print("Value: ", synthesizer.best_assignment_value)
```

considering the `model` and `property` is obtained in the way you described using stormpy
